extensions/amp-cld-img/0.1/test/test-amp-cld-img.js

```diff
@@ -0,0 +1,324 @@
+ */
+
+import '../amp-cld-img';
+import {buildUrl} from '../utils';
+
+describes.realWin('amp-cld-img', {
+  amp: {
+    extensions: ['amp-cld-img'],
+  },
+}, env => {
+
+  let win;
+  let doc;
+
+  beforeEach(() => {
+    win = env.win;
+    doc = win.document;
+    const script = doc.createElement('script');
+    script.id = 'amp-cld-config';
+    script.type = 'application/json';
+    script.textContent = '{ "cloudName": "demo" }';
+    doc.head.appendChild(script);
+  });
+
+  function getCldImg(extraAttrs) {
+    const cld = doc.createElement('amp-cld-img');
+    cld.setAttribute('data-public-id', 'sample');
+    cld.setAttribute('layout', 'fixed');
+    cld.setAttribute('width', '100');
+    cld.setAttribute('height', '100');
+
+    if (extraAttrs) {
+      Object.keys(extraAttrs).forEach(key =>
+        cld.setAttribute(key, extraAttrs[key]));
+    }
+
+    doc.body.appendChild(cld);
+    return cld.build().then(() => cld.layoutCallback()).then(() => cld);
+  }
+
+  it('should use attributes to generate url', () => {
+    return getCldImg().then(cld => {
+      expect(cld.childNodes[0].getAttribute('src'))
+          .to.eql('https://res.cloudinary.com/demo/image/upload/c_fill,h_100,w_100/sample');
+    });
+  });
+  it('should use tag attributes over global config', () => {
+    return getCldImg({'cloud-name': 'test123', 'crop': 'mfit'}).then(cld => {
+      expect(cld.childNodes[0].getAttribute('src'))
+          .to.eql('https://res.cloudinary.com/test123/image/upload/c_mfit,h_100,w_100/sample');
+    });
+  });
+  it('should use step size to generate transformation', () => {
+    return getCldImg({'step-size': '150'}).then(cld => {
+      expect(cld.childNodes[0].getAttribute('src'))
+          .to.eql('https://res.cloudinary.com/demo/image/upload/c_fill,h_150,w_150/sample');
+    });
+  });
+  it('should use max size to generate transformation', () => {
+    return getCldImg({'max-size': '60'}).then(cld => {
+      expect(cld.childNodes[0].getAttribute('src'))
+          .to.eql('https://res.cloudinary.com/demo/image/upload/c_fill,h_60,w_60/sample');
+    });
+  });
+  it('should use attributes to generate transformation', () => {
+    return getCldImg({'effect': 'blur:50'}).then(cld => {
+      expect(cld.childNodes[0].getAttribute('src'))
+          .to.eql('https://res.cloudinary.com/demo/image/upload/e_blur:50/c_fill,h_100,w_100/sample');
+    });
+  });
+  it('should use rawTransformation attribute', () => {
+    return getCldImg({'effect': 'blur:50',
+      'raw-transformation': 'l_sample,g_north'}).then(cld => {
+      expect(cld.childNodes[0].getAttribute('src'))
+          .to.eql('https://res.cloudinary.com/demo/image/upload/l_sample,g_north/e_blur:50/c_fill,h_100,w_100/sample');
+    });
+  });
+  it('should allow overriding cloudName in options', function() {
+    const options = {
+      cloudName: 'test321',
+    };
+    const result = buildUrl('test', options);
+    expect(options).to.eql({});
```

Comment:
Having an API that clears out an options argument doesn't seem very intuitive to me. If you need to remove keys from the options as you go along, you should first clone the input argument internally in the function.